### PR TITLE
Faster wasm runtime

### PR DESCRIPTION
**Pull request by benaadams**
* Compile the wasm using `-O3` rather than `-Oz` for a faster runtime
* Use Link Time Optimization `-flto` for extra optimizations
* Fix issues in JS and lay groundwork for closure compiler's `--closure 1` which can greatly reduce the size of the JS and more than make up for the increase in the wasm (see: https://github.com/dotnet/runtime/pull/51446)

File           | Branch | Uncompressed |     Brotli |
-------------- |-------:|-------------:|-----------:|
dotnet.wasm    |   main |     2,220 KB |     764 KB |
dotnet.wasm    |     PR |     2,338 KB |     771 KB |
 | | | |
**Net Change** |        |        +118 KB |     +7 KB |

[Optimizing Code | Link Times](https://emscripten.org/docs/optimizing/Optimizing-Code.html#link-times)

> Of course, for a final release build, it is usually worth linking with something like `-O3 --closure 1` for full optimizations.

[Trading off code size and performance](https://emscripten.org/docs/optimizing/Optimizing-Code.html#trading-off-code-size-and-performance)

> You may wish to build the less performance-sensitive source files in your project using -Os or -Oz and the remainder using -O2 (-Os and -Oz are similar to -O2, but reduce code size at the expense of performance. -Oz reduces code size more than -Os.)
>
> **Compiling with -Os or -Oz generally avoids inlining too.**

Inspired by [Is WebAssembly magic performance pixie dust?](https://surma.dev/things/js-to-asc/) where they were seeing a ore than x2 perf degradation from `-Os`:

> Another thing that the AssemblyScript folks pointed out to me is that the --optimize flag is equivalent to -O3s which aggressively optimizes for speed, but makes tradeoffs to reduce binary size. -O3 optimizes for speed and speed only. Having -O3s as a default is good in spirit — binary size matters on the web — but is it worth it? At least in this specific example the answer is no: -O3s ends up trading the laughable amount of ~30 bytes for a huge performance penalty

And its currently using `-Oz` which is even worse for performance than `-Os`

### CODE REVIEW

**Comment by benaadams** [comment]
@radekdoulik is it possible to test the performance of this with your example in https://github.com/dotnet/runtime/issues/50260#issuecomment-821173735; I'm not sure how to get a build that can be tested in your harness, I just seemed to make it angry in the console instead

File: src/mono/wasm/Makefile [hunk 1]
```diff
@@ -72,7 +72,7 @@ ifeq ($(ENABLE_METADATA_UPDATE),true)
 endif
 
 EMCC_DEBUG_FLAGS =-g -Os -s ASSERTIONS=1 -DDEBUG=1
-EMCC_RELEASE_FLAGS=-Oz --llvm-opts 2
+EMCC_RELEASE_FLAGS=-O3 --llvm-opts 3 -flto
```
Comment (eerhardt) [inline src/mono/wasm/Makefile:75]: (nit) not that it really matters, but the ordering here is different than what is in the wasm.proj file.

**Comment by eerhardt** [comment]
> 	$(EMCC) $(EMCC_FLAGS) $(1) --js-library runtime/library_mono.js --js-library runtime/binding_support.js --js-library runtime/dotnet_support.js --js-library $(SYSTEM_NATIVE_LIBDIR)/pal_random.js $(BUILDS_OBJ_DIR)/driver.o $(BUILDS_OBJ_DIR)/pinvoke.o $(BUILDS_OBJ_DIR)/corebindings.o $(2) -o $(NATIVE_BIN_DIR)/dotnet.js $(3)

Does this need to pass `--externs runtime/externs.js` on it?

---
Refers to: src/mono/wasm/Makefile:105 in 7ea1fa3. [](commit_id = 7ea1fa38b6b8dbf612578630e221d76227071623, deletion_comment = False)

**Comment by benaadams** [comment]
> Does this need to pass `--externs runtime/externs.js` on it?

That needs to be passed via `EMCC_CLOSURE_ARGS` environment var; so added that in the `RunWithEmSdkEnv.cs` Task, doesn't seem to be a simple way to add an env var in vanilla MsBuild?

**Comment by eerhardt** [comment]
> so added that in the RunWithEmSdkEnv.cs Task

I don't believe that Task gets called on non-Windows builds:

https://github.com/dotnet/runtime/blob/8f4a12df829ae1bd3ac8c39aa07730ebe286c849/src/mono/wasm/wasm.proj#L112-L115

vs.
https://github.com/dotnet/runtime/blob/8f4a12df829ae1bd3ac8c39aa07730ebe286c849/src/mono/wasm/wasm.proj#L74-L80

From my understanding the Makefile and the `BuildWinWasmRuntimes` target are duplicates of each other. cc @radekdoulik

**Comment by benaadams** [comment]
> I don't believe that Task gets called on non-Windows builds:

Added to the Makefile also

**Comment by radekdoulik** [comment]
> From my understanding the Makefile and the BuildWinWasmRuntimes target are duplicates of each other. cc @radekdoulik

Yes, indeed. I plan to unify that, hopefully soon. Opened https://github.com/dotnet/runtime/issues/51553.

**Review by radekdoulik — COMMENTED** [review]
I have measured the Json timings.

| measurement | main/interp | PR branch/interp |
|-:|-:|-:|
|   Json, non-ASCII text serialize |     8.2714ms |     8.2941ms |
| Json, non-ASCII text deserialize |    11.6825ms |    11.6508ms |
|            Json, small serialize |     0.2389ms |     0.2436ms |
|          Json, small deserialize |     0.3678ms |     0.3731ms |
|            Json, large serialize |    67.8831ms |    69.4267ms |
|          Json, large deserialize |   101.1538ms |   102.4038ms |

They are very close, probably just measurement errors. Not sure whether the change is too small or whether the Json [de]serialization is not affected by changes in this PR.

Any suggestion for an area where it might show larger difference?

**Comment by jeromelaban** [comment]
This change might have more impact with AOT enabled (`-flto` does, not sure about -O3). Note that `-flto` also has a very large impact on build time if not `-flto=thin` at link time.

I'm also wondering if this may help more the GC itself, what happens if you make lots of allocations during benchmarks?

**Comment by radekdoulik** [comment]
I think for impact on AOT we would need to also modify `src/mono/wasm/build` targets?

**Comment by radekdoulik** [comment]
I have opened draft PR with the simple benchmark sample I used to measure the times https://github.com/dotnet/runtime/pull/51559

**Comment by steveisok** [comment]
@lewing @radical do you think this is something we can/want to take?

**Comment by radekdoulik** [comment]
I am sorry @benaadams that this fall off the radar. We switched to `-O2` meanwhile for release builds and the build targets are also finally unified for win and linux/mac.

@radical, I think the flags can be overridden with `EmccLinkOptimizationFlag` property, right?

**Pull request closed without merging**